### PR TITLE
Add optionals 356

### LINKED ISSUE #356: Add Optionals
Where a null is used as a return type, change this to use an `Optional` instead. We don't want to use this for parameters

This is in line with a change to our style guide.

Branch: `0.2.0`

### PR DESCRIPTION
**Pull request by raulsoledispa**
## Summary of change

**Optional** is added inside the methods that return **null**

## Related issue

Closes #356 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-h6vx3n8l-KN8QnO50r7QWZHgHFnFdPw)

### CODE REVIEW

**Comment by knjk04** [comment]
Thanks @raulsoledispa for working on this. 

I'll mark this as a draft to give you time to complete the PR template. 

Also, as is stands, you won't show up as a contributor. One possible reason for this is because your email address attached to your commits is/are not the same as those associated with your GitHub account. 

In line with the ['Ensure your email address on your commits are correct'](https://github.com/Project-Books/book-project/blob/master/CONTRIBUTING.md#ensure-your-email-address-on-your-commits-are-correct) section our style guide, you'll need to fix this. Instructions on how to do so can be found [here](https://github.com/Project-Books/book-project/blob/master/CONTRIBUTING.md#rewriting-pushed-commits-with-a-different-email-address)

For more information, see https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-your-github-profile/why-are-my-contributions-not-showing-up-on-my-profile

I'll mark this PR as a draft. Feel free to request a review when you have fixed this.

If you need a hand with this, let us know over [Slack](https://join.slack.com/t/teambookproject/shared_invite/zt-h6vx3n8l-KN8QnO50r7QWZHgHFnFdPw).

**Comment by raulsoledispa** [comment]
@knjk04 Please to review of PR

**Comment by knjk04** [comment]
@raulsoledispa Thanks, we'll review shortly

File: src/main/java/com/karankumar/bookproject/backend/entity/RatingScale.java [hunk 1]
```diff
@@ -62,13 +66,13 @@ public String toString() {
...
+    public static Optional<Double> toDouble(RatingScale ratingScale) {
         if (ratingScale == null) {
-            LOGGER.log(Level.INFO, "Returning null as rating was null");
-            return null;
+            LOGGER.log(Level.INFO, "Returning optional as rating was null");
+            return empty();
         }
 
-        return ratingScale.value;
+        return ofNullable(ratingScale.value);
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/entity/RatingScale.java:75]: I guess this method can still return double and if scale is null we can return 0.0. I can see we are calling this method at lots of places and assigning a default value of 0.0

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 2]
```diff
@@ -355,9 +358,11 @@ private boolean isMovedToDifferentShelf() {
...
 
     private void setBookBean() {
-        Book book = populateBookBean();
-        if (book != null) {
-            binder.setBean(book);
+
+
+        Optional<Book> book = populateBookBean();
+        if (book.isPresent()) {
+            binder.setBean(book.get());
```
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:365]: Can be replaced with:
`populateBookBean().ifPresent(value -> binder.setBean(value));`

**Review by amit1307 — CHANGES_REQUESTED** [review]
There are lots of style guide violations like imports with * etc. Can you please refer to style guide and make sure changes are in line with the guide.
https://github.com/Project-Books/book-project/blob/master/STYLEGUIDE.md

File: src/main/java/com/karankumar/bookproject/backend/entity/RatingScale.java (hunk 1 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/entity/RatingScale.java:75]: > if scale is null we can return 0.0

On our [rating scale](https://github.com/Project-Books/book-project/blob/master/src/main/java/com/karankumar/bookproject/backend/entity/RatingScale.java#L29), we have 0.0, but we also have an enum value for `NO_RATING`, which should be the default

File: src/main/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtils.java [hunk 3]
```diff
@@ -18,12 +18,14 @@
 package com.karankumar.bookproject.backend.utils;
 
 import static com.karankumar.bookproject.backend.utils.ShelfUtils.isAllBooksShelf;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
 
 import java.util.Arrays;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Optional;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtils.java:27]: I've removed the wildcard import

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 4]
```diff
@@ -355,9 +358,11 @@ private boolean isMovedToDifferentShelf() {
     }
 
     private void setBookBean() {
-        Book book = populateBookBean();
-        if (book != null) {
-            binder.setBean(book);
+
+
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:362]: These lines can be removed

File: src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java [hunk 5]
```diff
@@ -74,22 +76,22 @@ void onlyReadBooksCountTowardsMostPagesStatistics() {
     @Test
     void testAveragePageLengthDivideByZero() {
         resetPageStatistics();
-        assertThat(pageStatistics.calculateAveragePageLength()).isNull();
+        assertTrue(pageStatistics.calculateAveragePageLength().isEmpty());
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java:79]: In line with our style guide, could you please replace this with `assertThat`? We're using AssertJ assertions

File: src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java [hunk 6]
```diff
@@ -74,22 +76,22 @@ void onlyReadBooksCountTowardsMostPagesStatistics() {
...
     }
 
     @Test
     void averagePageLengthExistsAndIsCorrect() {
         int averagePageLength =
                 StatisticTestUtils.getTotalNumberOfPages() / StatisticTestUtils.getNumberOfBooks();
-        assertThat(pageStatistics.calculateAveragePageLength()).isEqualTo(averagePageLength);
+        assertEquals(averagePageLength, pageStatistics.calculateAveragePageLength().orElse(0.0));
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java:86]: assertThat

File: src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java [hunk 7]
```diff
@@ -74,22 +76,22 @@ void onlyReadBooksCountTowardsMostPagesStatistics() {
...
     }
 
     @Test
     void testAveragePageLengthWithFloatPointCalculation() {
         StatisticTestUtils.deleteBook(StatisticTestUtils.getBookWithHighestRating());
         pageStatistics = new PageStatistics(predefinedShelfService);
         Double averagePageLength = 267.0;
-        assertThat(pageStatistics.calculateAveragePageLength()).isEqualTo(averagePageLength);
+        assertEquals(averagePageLength, pageStatistics.calculateAveragePageLength().orElse(0.0));
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/PageStatisticsTest.java:94]: assertThat

File: src/test/java/com/karankumar/bookproject/backend/statistics/RatingStatisticsTest.java [hunk 8]
```diff
@@ -56,42 +60,47 @@ public void beforeEachSetup() {
...
+                .orElse("");
         String expectedTitle = bookWithNoRating.getTitle();
         assertThat(actualTitle).isEqualTo(expectedTitle);
     }
 
     @Test
     void testNonExistentLowestRatedBook() {
         resetRatingStatistics();
-        assertThat(ratingStatistics.findLeastLikedBook()).isNull();
+        assertTrue(ratingStatistics.findLeastLikedBook().isEmpty());
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/RatingStatisticsTest.java:73]: assertThat

File: src/test/java/com/karankumar/bookproject/backend/statistics/RatingStatisticsTest.java [hunk 9]
```diff
@@ -56,42 +60,47 @@ public void beforeEachSetup() {
...
+                .orElse("");
         String expectedTitle = bookWithHighestRating.getTitle();
         assertThat(actualTitle).isEqualTo(expectedTitle);
     }
 
     @Test
     void testNonExistentHighestRatedBook() {
         resetRatingStatistics();
-        assertThat(ratingStatistics.findMostLikedBook()).isNull();
+        assertTrue(ratingStatistics.findMostLikedBook().isEmpty());
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/RatingStatisticsTest.java:89]: assertThat

File: src/test/java/com/karankumar/bookproject/backend/statistics/RatingStatisticsTest.java [hunk 10]
```diff
@@ -56,42 +60,47 @@ public void beforeEachSetup() {
...
         double average = totalRating / numberOfBooks;
-        assertThat(ratingStatistics.calculateAverageRatingGiven()).isEqualTo(average);
+        assertEquals(average, ratingStatistics.calculateAverageRatingGiven().orElse(0.0));
     }
 
     @Test
     void testAverageRatingDivideByZero() {
         resetRatingStatistics();
-        assertThat(ratingStatistics.calculateAverageRatingGiven()).isNull();
+        assertTrue(ratingStatistics.calculateAverageRatingGiven().isEmpty());
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/statistics/RatingStatisticsTest.java:103]: assertThat

File: src/test/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtilsTest.java [hunk 11]
```diff
@@ -239,9 +240,7 @@ void testGetPredefinedShelfNameReturnsCorrectShelf(String shelfName) {
             case "Did not finish":
                 expectedShelf = DID_NOT_FINISH;
         }
-        PredefinedShelf.ShelfName actualShelf =
-                predefinedShelfUtils.getPredefinedShelfName(shelfName);
-        assertThat(actualShelf).isEqualTo(expectedShelf);
+      assertEquals(expectedShelf, predefinedShelfUtils.getPredefinedShelfName(shelfName).orElse(PredefinedShelf.ShelfName.DID_NOT_FINISH));
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/utils/PredefinedShelfUtilsTest.java:243]: A couple of changes are needed here:

- assertThat
- Adhere to the 100 character column limit

**Review by knjk04 — CHANGES_REQUESTED** [review]
Thanks for working on this! 

I've left some feedback for you to address before we can merge

**Comment by knjk04** [comment]
Also, could you please resolve the conflicts and then rebase to ensure your branch is level with master?

**Comment by knjk04** [comment]
@raulsoledispa Do you have an update on the progress of addressing the review feedback?

If you need a hand with this, feel free to get in touch with us over our [Slack](https://join.slack.com/t/teambookproject/shared_invite/zt-i7lept44-rgJ9yB0A2vedJTLyyfkjKQ) workspace

**Comment by knjk04** [comment]
@raulsoledispa As we have not heard back from you in a while, despite following up, I'll close this pull request

**Pull request closed without merging**